File: src/engine/mcts.ts
/**
 * Information Set Monte Carlo Tree Search (ISMCTS) for Hanabi.
 *
 * Handles partial observability via determinization: sample possible hands
 * for the acting player consistent with their hint info, then run UCT search
 * with heuristic rollouts. One shared tree across all determinizations.
 *
 * Uses the tier-based action scoring from ai.ts:
 * - "bad" actions are filtered out
 * - "must" and "strong" actions are explored before "neutral"
 */
import type { GameState, Action, Card, Suit, Rank, Player } from './types';
import { SUITS, RANKS } from './types';
import { RANK_COUNTS } from './deck';
import { applyAction, getValidActions, getScore } from './game';
import { heuristicBotAction, scoreAction, TIER_ORDER } from './ai';
import type { ActionTier } from './ai';

const TIME_BUDGET_MS = 5000;
const EXPLORATION_C = 1.0;

// ---------------------------------------------------------------------------
// Tree node
// ---------------------------------------------------------------------------

class MCTSNode {
  parent: MCTSNode | null;
  action: Action | null;
  children: Map<string, MCTSNode> = new Map();
  visits = 0;
  totalScore = 0;

  constructor(parent: MCTSNode | null, action: Action | null) {
    this.parent = parent;
    this.action = action;
  }
}

// ---------------------------------------------------------------------------
// Action identity key (for shared tree across determinizations)
// ---------------------------------------------------------------------------

function actionKey(action: Action): string {
  switch (action.type) {
    case 'PLAY_CARD':
      return `P${action.cardIndex}`;
    case 'DISCARD':
      return `D${action.cardIndex}`;
    case 'GIVE_HINT': {
      const h = action.hint;
      const hk = h.kind === 'suit' ? `s${h.suit}` : `r${h.rank}`;
      return `H${action.targetPlayerIndex}${hk}`;
    }
  }
}

// ---------------------------------------------------------------------------
// State cloning (deep enough for simulation — cards are never mutated)
// ---------------------------------------------------------------------------

function cloneGameState(state: GameState): GameState {
  return {
    ...state,
    players: state.players.map(p => ({
      ...p,
      hand: [...p.hand],
      hintInfo: {
        knownSuits: [...p.hintInfo.knownSuits],
        knownRanks: [...p.hintInfo.knownRanks],
        cardAges: [...p.hintInfo.cardAges],
      },
    })),
    deck: [...state.deck],
    playArea: { ...state.playArea },
    discardPile: [...state.discardPile],
    history: [...state.history],
  };
}

// ---------------------------------------------------------------------------
// Determinization: sample unknown cards for the acting player
// ---------------------------------------------------------------------------

function shuffle<T>(arr: T[]): T[] {
  for (let i = arr.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [arr[i], arr[j]] = [arr[j], arr[i]];
  }
  return arr;
}

function buildUnknownPool(state: GameState, playerIndex: number): Card[] {
  const counts = new Map<string, number>();
  for (const suit of SUITS) {
    for (const rank of RANKS) {
      counts.set(`${suit}-${rank}`, RANK_COUNTS[rank]);
    }
  }

  for (const suit of SUITS) {
    for (let r = 1; r <= state.playArea[suit]; r++) {
      const key = `${suit}-${r}`;
      counts.set(key, counts.get(key)! - 1);
    }
  }

  for (const card of state.discardPile) {
    const key = `${card.suit}-${card.rank}`;
    counts.set(key, counts.get(key)! - 1);
  }

  for (let i = 0; i < state.players.length; i++) {
    if (i === playerIndex) continue;
    for (const card of state.players[i].hand) {
      const key = `${card.suit}-${card.rank}`;
      counts.set(key, counts.get(key)! - 1);
    }
  }

  const pool: Card[] = [];
  let idCounter = 0;
  for (const [key, count] of counts) {
    const [suit, rankStr] = key.split('-');
    for (let i = 0; i < count; i++) {
      pool.push({
        id: `unk-${idCounter++}`,
        suit: suit as Suit,
        rank: Number(rankStr) as Rank,
      });
    }
  }

  return pool;
}

function assignConsistentHand(
  pool: Card[],
  player: Player,
): { hand: Card[]; remaining: Card[] } | null {
  const handSize = player.hand.length;
  const available = [...pool];
  const hand: Card[] = [];

  for (let i = 0; i < handSize; i++) {
    const knownSuit = player.hintInfo.knownSuits[i];
    const knownRank = player.hintInfo.knownRanks[i];

    const candidates = available.filter(
      c =>
        (knownSuit === null || c.suit === knownSuit) &&
        (knownRank === null || c.rank === knownRank),
    );

    if (candidates.length === 0) return null;

    const pick = candidates[Math.floor(Math.random() * candidates.length)];
    hand.push(pick);
    const idx = available.indexOf(pick);
    available.splice(idx, 1);
  }

  return { hand, remaining: available };
}

function sampleDeterminization(state: GameState, playerIndex: number): GameState | null {
  const clone = cloneGameState(state);
  const pool = buildUnknownPool(state, playerIndex);
  shuffle(pool);

  const result = assignConsistentHand(pool, state.players[playerIndex]);
  if (!result) return null;

  clone.players[playerIndex] = {
    ...clone.players[playerIndex],
    hand: result.hand,
  };
  clone.deck = shuffle(result.remaining);

  return clone;
}

// ---------------------------------------------------------------------------
// Tier-based action filtering (replaces pruneActions)
// ---------------------------------------------------------------------------

interface RankedAction {
  action: Action;
  tier: ActionTier;
  score: number;
}

function filterAndRankActions(actions: Action[], state: GameState, playerIndex: number): RankedAction[] {
  const scored: RankedAction[] = actions.map(a => {
    const s = scoreAction(state, playerIndex, a);
    return { action: a, tier: s.tier, score: s.score };
  });

  // Filter out "bad" tier
  const filtered = scored.filter(s => s.tier !== 'bad');

  // Sort by tier (must > strong > neutral) then score descending
  filtered.sort((a, b) => {
    const tierDiff = TIER_ORDER[b.tier] - TIER_ORDER[a.tier];
    if (tierDiff !== 0) return tierDiff;
    return b.score - a.score;
  });

  // Safety: never filter all actions
  return filtered.length > 0 ? filtered : scored;
}

// ---------------------------------------------------------------------------
// UCT tree policy
// ---------------------------------------------------------------------------

function uctValue(child: MCTSNode, parentVisits: number): number {
  if (child.visits === 0) return Infinity;
  return (
    child.totalScore / child.visits +
    EXPLORATION_C * Math.sqrt(Math.log(parentVisits) / child.visits)
  );
}

function treeSelect(
  root: MCTSNode,
  state: GameState,
  rankedRootActions: RankedAction[],
  _rootPlayer: number,
): { node: MCTSNode; state: GameState; isRoot: boolean } {
  let node = root;
  let simState = state;
  let isRoot = true;

  while (simState.status === 'playing') {
    const actions = isRoot
      ? rankedRootActions
          .map(r => r.action)
          .filter(a => getValidActions(simState, simState.currentPlayerIndex).some(
            va => actionKey(va) === actionKey(a)))
      : getValidActions(simState, simState.currentPlayerIndex);

    if (actions.length === 0) break;

    const untriedActions = actions.filter(a => !node.children.has(actionKey(a)));
    if (untriedActions.length > 0) break;

    let bestChild: MCTSNode | null = null;
    let bestValue = -Infinity;

    for (const a of actions) {
      const key = actionKey(a);
      const child = node.children.get(key);
      if (!child) continue;
      const value = uctValue(child, node.visits);
      if (value > bestValue) {
        bestValue = value;
        bestChild = child;
      }
    }

    if (!bestChild) break;

    node = bestChild;
    simState = applyAction(simState, bestChild.action!);
    isRoot = false;
  }

  return { node, state: simState, isRoot };
}

// ---------------------------------------------------------------------------
// Expand: add one untried child
// At root: biased expansion (try highest-tier actions first)
// At deeper nodes: random expansion
// ---------------------------------------------------------------------------

function expand(
  node: MCTSNode,
  state: GameState,
  rankedRootActions: RankedAction[] | null,
  _rootPlayer: number,
): { node: MCTSNode; state: GameState } {
  if (state.status !== 'playing') return { node, state };

  const allActions = getValidActions(state, state.currentPlayerIndex);

  let untriedActions: Action[];

  if (rankedRootActions) {
    // Root level: use ranked actions, biased order (already sorted by tier+score)
    const validRanked = rankedRootActions
      .map(r => r.action)
      .filter(a => allActions.some(va => actionKey(va) === actionKey(a)));
    untriedActions = validRanked.filter(a => !node.children.has(actionKey(a)));
  } else {
    // Deeper nodes: random selection
    untriedActions = allActions.filter(a => !node.children.has(actionKey(a)));
  }

  if (untriedActions.length === 0) return { node, state };

  // At root: pick the first untried (highest-tier/score due to sorting)
  // At deeper nodes: pick randomly
  const action = rankedRootActions
    ? untriedActions[0]
    : untriedActions[Math.floor(Math.random() * untriedActions.length)];

  const child = new MCTSNode(node, action);
  node.children.set(actionKey(action), child);

  return { node: child, state: applyAction(state, action) };
}

// ---------------------------------------------------------------------------
// Rollout: play to completion using heuristic
// ---------------------------------------------------------------------------

function rollout(state: GameState): number {
  let simState = state;
  while (simState.status === 'playing') {
    const action = heuristicBotAction(simState, simState.currentPlayerIndex);
    simState = applyAction(simState, action);
  }
  return getScore(simState.playArea) / 25;
}

// ---------------------------------------------------------------------------
// Backpropagate
// ---------------------------------------------------------------------------

function backpropagate(node: MCTSNode | null, score: number): void {
  while (node !== null) {
    node.visits++;
    node.totalScore += score;
    node = node.parent;
  }
}

// ---------------------------------------------------------------------------
// Main entry point
// ---------------------------------------------------------------------------

const YIELD_INTERVAL_MS = 50;

function yieldToEventLoop(): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, 0));
}

export async function chooseMCTSAction(state: GameState, playerIndex: number): Promise<Action> {
  const allActions = getValidActions(state, playerIndex);
  const ranked = filterAndRankActions(allActions, state, playerIndex);

  // Fast path: if exactly one "must" action, take it immediately
  const mustActions = ranked.filter(r => r.tier === 'must');
  if (mustActions.length === 1) return mustActions[0].action;

  // If only one action survives filtering, just do it
  if (ranked.length === 1) return ranked[0].action;

  const root = new MCTSNode(null, null);
  const deadline = performance.now() + TIME_BUDGET_MS;
  let lastYield = performance.now();

  while (performance.now() < deadline) {
    if (performance.now() - lastYield > YIELD_INTERVAL_MS) {
      await yieldToEventLoop();
      lastYield = performance.now();
    }

    // 1. Determinize
    const det = sampleDeterminization(state, playerIndex);
    if (!det) continue;

    // 2. Select (using ranked actions at root level)
    const selected = treeSelect(root, det, ranked, playerIndex);

    // 3. Expand (biased at root, random at deeper nodes)
    const expanded = expand(selected.node, selected.state, selected.isRoot ? ranked : null, playerIndex);

    // 4. Rollout
    const score = rollout(expanded.state);

    // 5. Backpropagate
    backpropagate(expanded.node, score);
  }

  // Pick action with highest visit count
  let bestAction: Action | null = null;
  let bestVisits = -1;

  for (const [, child] of root.children) {
    if (child.visits > bestVisits) {
      bestVisits = child.visits;
      bestAction = child.action;
    }
  }

  if (!bestAction) {
    return heuristicBotAction(state, playerIndex);
  }

  // Safety: verify the chosen action is still valid in the real state
  if (!allActions.some(a => actionKey(a) === actionKey(bestAction!))) {
    return heuristicBotAction(state, playerIndex);
  }

  return bestAction;
}

File: training/mcts.py
"""
Information Set Monte Carlo Tree Search (ISMCTS) for Hanabi — Python port.

Used for ExIt data generation: run MCTS with neural rollout, collect
visit distributions at the root as soft training targets.
"""
import math
import time
import random
from copy import deepcopy

from engine import (
    SUITS, RANKS, RANK_COPIES, NUM_PLAYERS, HAND_SIZE,
    create_empty_hint_info, apply_action, get_valid_actions, get_score, clone_state
)


# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------

EXPLORATION_C = 1.0


# ---------------------------------------------------------------------------
# Tree node
# ---------------------------------------------------------------------------

class MCTSNode:
    __slots__ = ['parent', 'action_key', 'action', 'children', 'visits', 'total_score']

    def __init__(self, parent=None, action_key=None, action=None):
        self.parent = parent
        self.action_key = action_key
        self.action = action  # the game action dict that led to this node
        self.children = {}    # action_key -> MCTSNode
        self.visits = 0
        self.total_score = 0.0


# ---------------------------------------------------------------------------
# Action key (for shared tree across determinizations)
# ---------------------------------------------------------------------------

def action_key(action):
    atype = action['type']
    if atype == 'play':
        return f"P{action['card_idx']}"
    elif atype == 'discard':
        return f"D{action['card_idx']}"
    elif atype == 'hint':
        h = action['hint']
        hk = f"s{h['value']}" if h['kind'] == 'suit' else f"r{h['value']}"
        return f"H{action['target']}{hk}"


# ---------------------------------------------------------------------------
# Determinization: sample unknown cards for the acting player
# ---------------------------------------------------------------------------

def build_unknown_pool(state, player_idx):
    """Build pool of cards not visible to player_idx."""
    # Start with all 50 cards
    counts = {}
    for s in range(len(SUITS)):
        for r in RANKS:
            counts[(s, r)] = RANK_COPIES[r]

    # Subtract play area
    for s in range(len(SUITS)):
        for r in range(1, state['play_area'][s] + 1):
            counts[(s, r)] -= 1

    # Subtract discard pile
    for card in state['discard_pile']:
        counts[(card['suit'], card['rank'])] -= 1

    # Subtract other players' hands (visible)
    for i in range(NUM_PLAYERS):
        if i == player_idx:
            continue
        for card in state['players'][i]['hand']:
            counts[(card['suit'], card['rank'])] -= 1

    # Build pool
    pool = []
    for (s, r), count in counts.items():
        for _ in range(count):
            pool.append({'suit': s, 'rank': r})

    return pool


def assign_consistent_hand(pool, player):
    """Try to assign cards from pool consistent with player's hints."""
    hand_size = len(player['hand'])
    available = list(pool)
    hand = []

    for i in range(hand_size):
        ks = player['hint_info']['known_suits'][i]
        kr = player['hint_info']['known_ranks'][i]

        candidates = [c for c in available
                      if (ks is None or c['suit'] == ks)
                      and (kr is None or c['rank'] == kr)]

        if not candidates:
            return None  # rejection — retry

        pick = random.choice(candidates)
        hand.append(pick)
        available.remove(pick)

    return hand, available


def sample_determinization(state, player_idx):
    """Create a determinized copy where player_idx's hand is sampled."""
    clone = clone_state(state)
    pool = build_unknown_pool(state, player_idx)
    random.shuffle(pool)

    result = assign_consistent_hand(pool, state['players'][player_idx])
    if result is None:
        return None

    hand, remaining = result
    clone['players'][player_idx]['hand'] = hand
    # Keep hint info from original (it's about what the player knows, not the actual cards)
    random.shuffle(remaining)
    clone['deck'] = remaining

    return clone


# ---------------------------------------------------------------------------
# Action pruning (same logic as TypeScript mcts.ts)
# ---------------------------------------------------------------------------

def filter_and_rank_actions(actions, state, player_idx):
    """Filter out 'bad' tier actions and sort by tier+score (best first)."""
    from heuristic_bot import score_action, TIER_ORDER

    scored = []
    for a in actions:
        tier, sc = score_action(state, player_idx, a)
        scored.append((a, tier, sc))

    # Filter out bad tier
    filtered = [(a, t, s) for a, t, s in scored if t != 'bad']

    # Sort by tier desc, then score desc
    filtered.sort(key=lambda x: (TIER_ORDER[x[1]], x[2]), reverse=True)

    # Safety: never filter all actions
    if not filtered:
        scored.sort(key=lambda x: (TIER_ORDER[x[1]], x[2]), reverse=True)
        return [a for a, _, _ in scored]

    return [a for a, _, _ in filtered]


# ---------------------------------------------------------------------------
# UCT
# ---------------------------------------------------------------------------

def uct_value(child, parent_visits):
    if child.visits == 0:
        return float('inf')
    return (child.total_score / child.visits +
            EXPLORATION_C * math.sqrt(math.log(parent_visits) / child.visits))


def tree_select(root, state, pruned_root_actions):
    node = root
    sim_state = state
    is_root = True

    while sim_state['status'] == 'playing':
        if is_root:
            valid = get_valid_actions(sim_state, sim_state['current_player'])
            valid_keys = {action_key(a) for a in valid}
            actions = [a for a in pruned_root_actions if action_key(a) in valid_keys]
        else:
            actions = get_valid_actions(sim_state, sim_state['current_player'])

        if not actions:
            break

        untried = [a for a in actions if action_key(a) not in node.children]
        if untried:
            break

        best_child = None
        best_value = -float('inf')
        for a in actions:
            key = action_key(a)
            child = node.children.get(key)
            if child is None:
                continue
            val = uct_value(child, node.visits)
            if val > best_value:
                best_value = val
                best_child = child
                best_action = a

        if best_child is None:
            break

        node = best_child
        apply_action(sim_state, best_action)
        is_root = False

    return node, sim_state, is_root


def expand(node, state, ranked_root_actions):
    if state['status'] != 'playing':
        return node, state

    all_actions = get_valid_actions(state, state['current_player'])
    if ranked_root_actions is not None:
        # Root level: use ranked actions (already sorted by tier+score)
        valid_keys = {action_key(a) for a in all_actions}
        actions = [a for a in ranked_root_actions if action_key(a) in valid_keys]
    else:
        actions = all_actions

    untried = [a for a in actions if action_key(a) not in node.children]
    if not untried:
        return node, state

    # At root: pick first untried (biased toward best tier+score)
    # At deeper nodes: pick randomly
    action = untried[0] if ranked_root_actions is not None else random.choice(untried)
    key = action_key(action)
    child = MCTSNode(parent=node, action_key=key, action=action)
    node.children[key] = child

    apply_action(state, action)
    return child, state


def backpropagate(node, score):
    while node is not None:
        node.visits += 1
        node.total_score += score
        node = node.parent


# ---------------------------------------------------------------------------
# Rollout
# ---------------------------------------------------------------------------

def rollout_with_policy(state, policy_fn):
    """Play to completion using policy_fn(state, player_idx) -> action."""
    sim_state = state
    while sim_state['status'] == 'playing':
        action = policy_fn(sim_state, sim_state['current_player'])
        apply_action(sim_state, action)
    return get_score(sim_state) / 25.0


# ---------------------------------------------------------------------------
# Main MCTS
# ---------------------------------------------------------------------------

def run_mcts(state, player_idx, policy_fn, time_budget_ms=200):
    """Run MCTS and return visit distribution over actions.

    Args:
        state: game state (not mutated)
        player_idx: acting player
        policy_fn: rollout policy function(state, player_idx) -> action
        time_budget_ms: search budget in milliseconds

    Returns:
        List of (action, visit_count) tuples for root children.
    """
    all_actions = get_valid_actions(state, player_idx)
    if not all_actions:
        return []

    ranked = filter_and_rank_actions(all_actions, state, player_idx)
    if len(ranked) == 1:
        return [(ranked[0], 1)]

    root = MCTSNode()
    deadline = time.monotonic() + time_budget_ms / 1000.0
    sim_count = 0

    while time.monotonic() < deadline:
        # 1. Determinize
        det = sample_determinization(state, player_idx)
        if det is None:
            continue

        # 2. Select
        node, sim_state, is_root = tree_select(root, det, ranked)

        # 3. Expand (biased at root, random at deeper nodes)
        node, sim_state = expand(node, sim_state, ranked if is_root else None)

        # 4. Rollout
        score = rollout_with_policy(sim_state, policy_fn)

        # 5. Backpropagate
        backpropagate(node, score)
        sim_count += 1

    # Collect visit distribution
    results = []
    for key, child in root.children.items():
        if hasattr(child, 'action'):
            results.append((child.action, child.visits))

    return results
